File: java/bird-watcher/src/main/java/BirdWatcher.java

class BirdWatcher {
    private final int[] birdsPerDay;


    public BirdWatcher(int[] birdsPerDay) {
        this.birdsPerDay = birdsPerDay.clone();
    }

    public int[] getLastWeek() {
        return birdsPerDay;
    }

    public int getToday() {
        return birdsPerDay[birdsPerDay.length - 1];
    }

    public void incrementTodaysCount() {
        birdsPerDay[birdsPerDay.length - 1]++;
    }

    public boolean hasDayWithoutBirds() {
        boolean zeroDay = false;
        for (int dayNum : birdsPerDay) {
            if (dayNum == 0)
                zeroDay = true;
        }

        return zeroDay;
    }

    public int getCountForFirstDays(int numberOfDays) {
        int sum = 0;
        if (numberOfDays > birdsPerDay.length)
            numberOfDays = birdsPerDay.length;

        for (int i = 0; i < numberOfDays; i++) {
            sum += birdsPerDay[i];
        }

        return sum;
    }

    public int getBusyDays() {
        int total = 0;

        for (int day : birdsPerDay) {
            if (day >= 5) {
                total++;
            }
        }
        return total;
    }
}

File: fsharp/bird-watcher/BirdWatcher.fs
module BirdWatcher

open System.Linq

let lastWeek: int[] =
    [| 0; 2; 5; 3; 7; 8; 4 |]

// Yesterday is the 5th day
let yesterday(counts: int[]): int = counts[5]

let total(counts: int[]): int = counts.Sum()

let dayWithoutBirds(counts: int[]): bool =
    Array.contains 0 counts

let incrementTodaysCount(counts: int[]): int[] =
    let today = counts.Length - 1
    counts[today] <- counts[today] + 1
    counts

/// Every odd day has the same number of birds
let unusualWeek(counts: int[]): bool =
    match counts with
    | [| _; m; _; n; _; o; _ |] when m = n && n = o -> true
    | [| m; _; n; _; o; _; p |] when m = n && m = n && m = o && m = p -> true
    | _  -> false
